### PR TITLE
Upstream mmdetection 2.0

### PR DESCRIPTION
**Pull request by Ilya-Krylov**

### CODE REVIEW

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by zr526799544** [comment]
> run tests

It seem failed in mmdetv2.0. I find some thing diferent in AnchorGenerator.grid_anchors(). featmap_size in mmdetv2.0 is a torch.size. however. It become a tensor in your jit trace. could you give me some advice to fix it????

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by AlexanderDokuchaev** [comment]
run tests

**Comment by AlexanderDokuchaev** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

File: mmdet/core/anchor/anchor_generator.py [hunk 1]
```diff
@@ -102,7 +255,281 @@ def valid_flags(self, featmap_size, valid_size, device='cuda'):
...
+        repr_str += f'{indent_str}scales_per_octave='
+        repr_str += f'{self.scales_per_octave},\n'
+        repr_str += f'{indent_str}num_levels={self.num_levels}\n'
+        repr_str += f'{indent_str}centers={self.centers},\n'
+        repr_str += f'{indent_str}center_offset={self.center_offset})'
+        return repr_str
+
+
+@ANCHOR_GENERATORS.register_module()
+class SSDAnchorGenerator(AnchorGenerator):
```
Comment (druzhkov-paul) [inline mmdet/core/anchor/anchor_generator.py:281]: Is it only applicable to SSD models?

File: mmdet/models/roi_heads/standard_roi_head.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+                    x,
+                    proposal_list,
+                    img_metas,
+                    proposals=None,
+                    rescale=False,
+                    postprocess=True):
+        """Test without augmentation."""
+        assert self.with_bbox, 'Bbox head must be implemented.'
+
+        if 0:
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/standard_roi_head.py]: I believe this is a debug code. If so, please remove it.

**Review by druzhkov-paul — CHANGES_REQUESTED** [review]

**Comment by druzhkov-paul** [comment]
@Ilya-Krylov Looks good!

I'd propose to merge it as soon as the tests go green. This being said, I think some extra effort should be put here afterwards. Here're some of the points that may be addressed later:

- Verification of dependencies' versions. Cython, PyTorch and NumPy in particular.
- Better alignment of the docs.
- Revision of NMS code and its export.
- Revision of masks' postprocessing code.

I'll address some of these myself, don't treat it as an action request for now.

**Comment by Ilya-Krylov** [comment]
run tests

File: mmdet/core/anchor/anchor_generator.py (hunk 1 above)
Comment (Ilya-Krylov) [reply mmdet/core/anchor/anchor_generator.py:281]: hz - it is from upstream

File: mmdet/models/roi_heads/standard_roi_head.py (hunk 2 above)
Comment (Ilya-Krylov) [reply mmdet/models/roi_heads/standard_roi_head.py]: fixed

**Comment by Ilya-Krylov** [comment]
run tests

File: mmdet/core/anchor/anchor_generator.py (hunk 1 above)
Comment (druzhkov-paul) [reply mmdet/core/anchor/anchor_generator.py:281]: Don't see `SSDAnchorGenerator` in upstream. What am I missing?

Comment (Ilya-Krylov) [reply mmdet/core/anchor/anchor_generator.py:281]: It was in tag v2.0.0 https://github.com/open-mmlab/mmdetection/blob/v2.0.0/mmdet/core/anchor/anchor_generator.py

Comment (druzhkov-paul) [reply mmdet/core/anchor/anchor_generator.py:281]: Thanks. My bad.

**Review by druzhkov-paul — APPROVED** [review]

**Pull request merged**